mlflow/data/tensorflow_dataset.py

```diff
@@ -102,83 +127,120 @@ def profile(self) -> Optional[Any]:
         """
         A profile of the dataset. May be None if no profile is available.
         """
-        import tensorflow as tf
-
         profile = {
-            "features_num_rows": len(self._data),
-            "features_num_elements": int(self._data.cardinality().numpy())
-            if isinstance(self._data, tf.data.Dataset)
-            else int(tf.size(self._data).numpy()),
+            "features_cardinality": int(self._features.cardinality().numpy())
+            if isinstance(self._features, tf.data.Dataset)
+            else int(tf.size(self._features).numpy()),
         }
         if self._targets is not None:
             profile.update(
                 {
-                    "targets_num_rows": len(self._targets),
-                    "targets_num_elements": int(self._targets.cardinality().numpy())
+                    "targets_cardinality": int(self._targets.cardinality().numpy())
                     if isinstance(self._targets, tf.data.Dataset)
                     else int(tf.size(self._targets).numpy()),
                 }
             )
         return profile
 
     @cached_property
-    def schema(self) -> Optional[Schema]:
+    def schema(self) -> Optional[TensorDatasetSchema]:
         """
         An MLflow TensorSpec schema representing the tensor dataset
         """
-        import tensorflow as tf
-
         try:
-            if isinstance(self._data, tf.data.Dataset):
-                features_tensor = next(self._data.as_numpy_iterator())
-                if isinstance(features_tensor, tuple):
-                    features_tensor = features_tensor[0]
-            else:
-                features_tensor = self._data.numpy()
-
-            schema_dict = {"features": features_tensor}
+            features_schema = TensorflowDataset._get_tf_object_schema(self._features)
+            targets_schema = None
             if self._targets is not None:
-                if isinstance(self._targets, tf.data.Dataset):
-                    targets_tensor = next(self._targets.as_numpy_iterator())
-                    if isinstance(targets_tensor, tuple):
-                        targets_tensor = targets_tensor[0]
-                else:
-                    targets_tensor = self._targets.numpy()
-                schema_dict["targets"] = targets_tensor
-            return _infer_schema(schema_dict)
```

Comment:
This logic was incorrect, since:

- it ignores all feature tensors except the first tensor
- it ignores all target tensors except the first tensor